reactioncommerce/reaction
# Multi-shop permission fixes

Impact: **minor**  
Type: **bugfix**

## Issues
- "products/updateVariant" Meteor method had questionable security and was unused
- "products/updateVariantsPosition" Meteor method did not check permissions for correct shop. Someone with permission to change variant order for any shop could also do it for all other shops.
- "products/getpublishedProductHash" Meteor method was not secured and was unused
- "shop/changeLayouts" Meteor method was not secured and was unused
- "shop/fetchCurrencyRate" Meteor method was not secured. Anybody, even if not logged in, could trigger a query of the currency rate exchange API and subsequent update of `currencies` property of the primary shop.
- "shop/flushCurrencyRate" Meteor method was not secured. Anybody, even if not logged in, could trigger the clearing of all `rate` properties in the `currencies` property of the primary shop.
- "shop/hideHeaderTag" Meteor method did not properly check shopId and was unused.
- "shop/updateHeaderTags" Meteor method did not properly check shopId. If you knew a tag ID and had proper permissions for your own shop, you could change the name of that tag even if it belonged to a different shop.
- "shop/removeHeaderTag" Meteor method did not check permissions for correct shop and was unused.
- "shop/togglePackage" Meteor method did not check permissions for correct shop. If you had permission for your shop, you could toggle plugins enabled status for another shop.
- "registry/update" Meteor method did not check permissions for correct shop. If you had permission for your shop, you could edit plugin settings for another shop.

## Solutions
- "products/updateVariant" Meteor method is removed
- "products/updateVariantsPosition" Meteor method checks permissions for the correct shop.
- "products/getpublishedProductHash" Meteor method is removed
- "shop/changeLayouts" Meteor method is removed
- "shop/fetchCurrencyRate" Meteor method is removed. The code is now in a private server method because it is only called from background job workers.
- "shop/flushCurrencyRate" Meteor method is removed. The code is now in a private server method because it is only called from background job workers.
- "shop/hideHeaderTag" Meteor method is removed.
- "shop/updateHeaderTags" Meteor method now checks permissions for the correct shop.
- "shop/removeHeaderTag" Meteor method is removed.
- "shop/togglePackage" Meteor method now checks permissions for the correct shop.
- "registry/update" Meteor method now checks permissions for the correct shop.

## Breaking changes
Custom code relying on the removed Meteor methods will break.

## Testing
1. Verify that the removed Meteor methods can no longer be called from a connected browser.
2. Verify that the listed permission issues with the not-removed Meteor methods are fixed. You will need to create multiple shops to test some fixes, for example by enabling and using the Marketplace plugin.

To call Meteor methods from a browser, connect to localhost:3000, open the browser console, and enter `Meteor.call("methodName", /* all necessary args */, console.log.bind(console))`. By passing `console.log` as the callback function, it will be sure to log the results and/or error when it's done.
